library/Freckle/App/Stats.hs

```diff
@@ -111,14 +115,33 @@ withStatsClient StatsSettings {..} f = do
   if amsEnabled
     then do
       tags <- (amsTags <>) <$> getEcsMetadataTags
-
+      gauges <- liftIO $ newIORef mempty
       Datadog.withDogStatsD amsSettings $ \client ->
         -- Add the tags to the thread context so they're present in all logs
         withThreadContext (map toPair tags)
-          $ f StatsClient { scClient = client, scTags = tags }
-    else f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags }
+          $ f StatsClient { scClient = client, scTags = tags, scGauges = gauges }
+    else do
+      gauges <- liftIO $ newIORef mempty
+      f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags, scGauges = gauges }
   where toPair = bimap (fromString . unpack) String
 
+withGauge :: (MonadReader app m, HasStatsClient app, MonadUnliftIO m) => Text -> m a -> m a
+withGauge name f = do
+  gaugesRef <- asks $ scGauges . view statsClientL
+  gaugesMap <- liftIO $ readIORef gaugesRef
+  gaugeValue <- liftIO $ maybe (createNewEKG gaugesRef gaugesMap) pure (HashMap.lookup name gaugesMap)
+  bracketEKG f gaugeValue
+  where
+    createNewEKG ref hashmap = do
+      ekg <- EKG.new
+      liftIO $ writeIORef ref (HashMap.insert name ekg hashmap)
+      pure ekg
```

Comment:
There are a few ways to get rid of all the repeated `liftIO`s. The one I think I would use is to use `UnliftIO.IORef` which is already lifted.

```suggestion
  gaugesRef <- asks $ scGauges . view statsClientL
  gaugesMap <- readIORef gaugesRef
  gaugeValue <- maybe (createNewEKG gaugesRef gaugesMap) pure (HashMap.lookup name gaugesMap)
  bracketEKG f gaugeValue
  where
    createNewEKG ref hashmap = do
      ekg <- liftIO EKG.new
      ekg <$ writeIORef ref (HashMap.insert name ekg hashmap)
```